src/csharp/Microsoft.Spark/Broadcast.cs

```diff
@@ -0,0 +1,265 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.IO;
+using System.Runtime.Serialization.Formatters.Binary;
+using Microsoft.Spark.Interop.Ipc;
+using Microsoft.Spark.Interop.Internal.Java.Util;
+using Microsoft.Spark.Interop;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.Spark
+{
+    /// <summary>
+    /// A broadcast variable. Broadcast variables allow the programmer to keep a read-only variable
+    /// cached on each machine rather than shipping a copy of it with tasks. They can be used, for
+    /// example, to give every node a copy of a large input dataset in an efficient manner. Spark
+    /// also attempts to distribute broadcast variables using efficient broadcast algorithms to
+    /// reduce communication cost.
+    /// </summary>
+    [Serializable]
+    public sealed class Broadcast<T>: IJvmObjectReferenceProvider
+    {
+        [NonSerialized]
+        private readonly JvmObjectReference _jvmObject;
+        private readonly long _bid;
+        private string _path;
+
+        internal Broadcast(SparkContext sc,
+            object value,
+            string localDir,
+            JvmObjectReference sparkContext)
+        {
+            var tempDir = Path.Combine(localDir, "sparkdotnet");
+            _path = Path.Combine(tempDir, Path.GetRandomFileName());
+            Version version = SparkEnvironment.SparkVersion;
+
+            // For Spark versions 2.3.0 and 2.3.1, Broadcast variable is created through different
+            // functions.
+            _jvmObject = (version.Major, version.Minor) switch
+            {
+                (2, 3) => (version.Build) switch
+                {
+                    0 => CreateBroadcast_V2_3_1_Below(tempDir, sparkContext, value),
```

Comment:
Maybe just
```C#
            _jvmObject = (version.Major, version.Minor) switch
            {
                (2, 3) when version.Build == 0 || version.Build == 1 =>
                    CreateBroadcast_V2_3_1_Below(tempDir, sparkContext, value),
                (2, 3) => CreateBroadcast_V2_3_2_Above(sc, tempDir, sparkContext, value),
                (2, 4) => CreateBroadcast_V2_3_2_Above(sc, tempDir, sparkContext, value),
                (3, 0) => CreateBroadcast_V2_3_2_Above(sc, tempDir, sparkContext, value),
                _ => throw new NotSupportedException($"Spark {version} not supported.")
            };
```

Reply:
And let's not add 3.0 yet unless you tested against it.

Reply:
Makes sense, will do. Thanks!